mgmt: mcumgr: fs: Add checksum (CRC32), hash (SHA256) and file size handlers

nordicjm opened the pull request:
doc: release-notes: Add mcumgr file status, file hash/checksum details
    
    This adds release notes for the new file status and file hash/checksum
    mcumgr fs commands.
    
    doc: guides: device_mgmt: smp: fs: Add hash/checksum/status details
    
    This adds documentation which explains fs mcumgr commands for the status
    and hash/checksum commands.
    
    mgmt: mcumgr: fs: Add hash/checksum (with CRC32/SHA256) handler
    
    This adds a hash/checksum mcumgr handler to the file management commands
    which can be used to get a hask or checksum of a file, and includes
    handler implementations for IEEE CRC32 and SHA256.
    
    mgmt: mcumgr: fs: Add file status handler
    
    This adds a command handler to the file management mcumgr system to get
    the status of a file without needing to return file data (currently
    reporting the file size).

Commit:
mgmt: mcumgr: fs: Add hash/checksum (with CRC32/SHA256) handler

This adds a hash/checksum mcumgr handler to the file management commands
which can be used to get a hask or checksum of a file, and includes
handler implementations for IEEE CRC32 and SHA256.

Signed-off-by: Jamie McCrae <jamie.mccrae@lairdconnect.com>

de-nordic commented on subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/include/fs_mgmt/fs_mgmt.h:
```diff
@@ -14,7 +14,9 @@ extern "C" {
  */
-#define FS_MGMT_ID_FILE	 0
+#define FS_MGMT_ID_FILE		0
+#define FS_MGMT_ID_CRC32	1
+#define FS_MGMT_ID_SHA256	2
```
Do not add these as separate commands, use one command like `FS_MGMT_ID_CHKSUM` instead with param that would allow to select checksum algorithm. Otherwise we start to bloat IDs with all the requested checksums.

Commit:
mgmt: mcumgr: fs: Add file status handler

This adds a command handler to the file management mcumgr system to get
the status of a file without needing to return file data (currently
reporting the file size).

Signed-off-by: Jamie McCrae <jamie.mccrae@lairdconnect.com>

de-nordic commented on subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/fs_mgmt.c:
```diff
@@ -214,11 +215,205 @@ fs_mgmt_file_upload(struct mgmt_ctxt *ctxt)
+	}
+
+	/* Encode the response. */
+	err = 0;
+	err |= cbor_encode_text_stringz(&ctxt->encoder, "rc");
```
I see no point in assigning here 0, just assign result of first cbor call directly.

de-nordic commented on subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/fs_mgmt.c:
```diff
@@ -214,11 +215,205 @@ fs_mgmt_file_upload(struct mgmt_ctxt *ctxt)
+	if (rc != 0 || type >= FS_MGMT_HASH_COUNT) {
+		return MGMT_ERR_EINVAL;
+	}
+
```
I think that it is worth adding `path[ARRAY_SIZE(path) - 1] = '\0'` here to "just-in-case" make sure that null terminator is there.

de-nordic commented on subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/fs_mgmt.c:
```diff
@@ -214,11 +215,205 @@ fs_mgmt_file_upload(struct mgmt_ctxt *ctxt)
+	if (rc != 0 || type >= FS_MGMT_CHECKSUM_COUNT) {
+		return MGMT_ERR_EINVAL;
+	}
+
```
I think that it is worth adding `path[ARRAY_SIZE(path) - 1] = '\0'` here to "just-in-case" make sure that null terminator is there.

de-nordic commented on subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/fs_mgmt.c:
```diff
@@ -214,11 +215,205 @@ fs_mgmt_file_upload(struct mgmt_ctxt *ctxt)
 		.mh_read = fs_mgmt_file_download,
 		.mh_write = fs_mgmt_file_upload,
 	},
+	[FS_MGMT_ID_SIZE] = {
+		.mh_read = fs_mgmt_file_size,
+		.mh_write = NULL,
+	},
+	[FS_MGMT_ID_CHECKSUM] = {
+		.mh_read = fs_mgmt_file_checksum,
+		.mh_write = NULL,
+	},
+	[FS_MGMT_ID_HASH] = {
+		.mh_read = fs_mgmt_file_hash,
+		.mh_write = NULL,
+	},
```
When they are not supported, do not add them here: the mcumgr will respond with error 8 when unable to find handler, and adding these makes it just do sequence:
1) search handler
2) call handler
3) handler returns with error
4) reset SMP buffer and form "rc" response with error returned from handler

Instead of directly going:
1) search handler;
1) send "rc" response with error 8
Additionally this adds code, when commands are not supported, that duplicates already existing behavior of SMP request processor.

de-nordic commented on subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/zephyr_fs_mgmt.c:
```diff
@@ -28,9 +42,7 @@ fs_mgmt_impl_filelen(const char *path, size_t *out_len)
-int
-fs_mgmt_impl_read(const char *path, size_t offset, size_t len,
-				  void *out_data, size_t *out_len)
+int fs_mgmt_impl_read(const char *path, size_t offset, size_t len, void *out_data, size_t *out_len)
```
Style changes like this, the `zephyr_fs_mgmt_truncate` and `zephyr_fs_mgmt_truncate` should at least get separate commit, or PR. Otherwise this just bloats PR diff with changes that are not related to the subject of the change.

lairdjm replied:
This was caused by using clang format to try and get rid of the 'do...while' error, should I just revert the changes and ignore that error?

de-nordic reviewed (changes requested):
Aside from comments, also needed:
1) release notes entry on new Kconfigs and commands added to mcumgr;
2) SMP documentation update, probably - new commands are added.

de-nordic commented:
> @de-nordic Does it make sense to have an additional function which returns an array/map of the supported checksum/hash functions on a device?

No and yes, it depends: In most cases you would know what sums you have implemented and what you want to request, on the other side you may have some devices that have newer software and support additional sums and you may query the device to know what sums you may use.

Log mgmt/mcumgr/lib Feature request for this so we can focus on having the current PR merged.

Commit:
doc: guides: device_mgmt: smp: fs: Add hash/checksum/status details

This adds documentation which explains fs mcumgr commands for the status
and hash/checksum commands.

Signed-off-by: Jamie McCrae <jamie.mccrae@lairdconnect.com>

Commit:
doc: release-notes: Add mcumgr file status, file hash/checksum details

This adds release notes for the new file status and file hash/checksum
mcumgr fs commands.

Signed-off-by: Jamie McCrae <jamie.mccrae@lairdconnect.com>

lairdjm commented:
Added length and offset parameters

nvlsianpu reviewed (commented):
@de-nordic As I understand you are accepting addition of functionalities in this PR. Am I right?

de-nordic commented on doc/guides/device_mgmt/smp_groups/smp_group_8.rst:
```diff
@@ -216,3 +220,209 @@ where:
+
+    {
+        (str)"len"      : (uint)
+        (str)"rc"       : (int)
```
Do the rc optional; the rc should only be sent when there is an error.

de-nordic commented on doc/guides/device_mgmt/smp_groups/smp_group_8.rst:
```diff
@@ -216,3 +220,209 @@ where:
+.. code-block:: none
+
+    {
+        (str)"type"     : (str)
```
Make type optional. The type is returned in response anyway and very often users will only rely on whatever default type would be set.

lairdjm replied:
Done, made type selection automatic however.

de-nordic commented on subsys/mgmt/mcumgr/Kconfig:
```diff
@@ -131,6 +131,56 @@ config FS_MGMT_DL_CHUNK_SIZE
 
 endif
 
+config FS_MGMT_FILE_STATUS_ENABLED
```
`_ENABLED` non needed. For example look at this `IS_ENABLED(CONFIG_FS_MGMT_FILE_STATUS_ENABLED)` would look strange.

lairdjm replied:
Done

de-nordic commented on subsys/mgmt/mcumgr/Kconfig:
```diff
@@ -131,6 +131,56 @@ config FS_MGMT_DL_CHUNK_SIZE
+	  This command allows a remote device to retrieve the status of a file,
+	  at present only the size of the file is returned (if it exists).
+
+config FS_MGMT_CHECKSUM_HASH_SUPPORTED
```
Just `FS_MGMT_CHECKSUM_HASH`

lairdjm replied:
Done

de-nordic commented on subsys/mgmt/mcumgr/Kconfig:
```diff
@@ -131,6 +131,56 @@ config FS_MGMT_DL_CHUNK_SIZE
+
+config FS_MGMT_CHECKSUM_HASH_SUPPORTED
+	bool "Checksum/hash mcumgr functions"
+	depends on HEAP_MEM_POOL_SIZE != 0
```
This is very weak dependency, you can set the `HEAP_MEM_POOL_SIZE` to any non-0 and the functions will still not work because something else will take all of the heap.

lairdjm replied:
Changed to stack, default at 128

de-nordic commented on subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/fs_mgmt.c:
```diff
@@ -208,11 +226,235 @@ fs_mgmt_file_upload(struct mgmt_ctxt *ctxt)
+	}
+
+	/* Allocate a buffer for the output data */
+	output = k_malloc(group->output_size);
```
I do not think that it is worth to engage heap for this. Just define buffer that will take the longest currently supported hash, as selected by Kconfig, either on stack or as static buffer.

lairdjm replied:
Uses longest supported hash/checksum size for buffer

de-nordic commented on subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/fs_mgmt.c:
```diff
@@ -208,11 +226,235 @@ fs_mgmt_file_upload(struct mgmt_ctxt *ctxt)
+		return MGMT_ERR_ENOENT;
+	}
+
+	if (file_len <= off) {
```
Shouldn't this be
```
if (file_len < off + len)
```
?

lairdjm replied:
`file_len` is the maximum size to use, so if the file size is not known then the maximum value would be accepted rather than having to query the file size first and then getting a hash/checksum.

de-nordic commented on subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/hash_checksum_crc32.c:
```diff
@@ -0,0 +1,66 @@
+		crc32 = crc32_ieee_update(crc32, buffer, bytes_read);
+		*out_len += bytes_read;
+
+		if (*out_len >= len) {
```
This can be incorporated into `while` loop condition.

lairdjm replied:
Done

de-nordic commented on subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/hash_checksum_crc32.c:
```diff
@@ -0,0 +1,66 @@
+
+	memcpy(output, &crc32, sizeof(crc32));
+
+	return 0;
```
Shouldn't we return error here when `bytes_read < 0`? User will only notice failure when it is know what is file size, otherwise this just returns crc for any number of bytes read and that size will be reported in response.

lairdjm replied:
I've added an error in but just realised it's a zephyr errno code rather than mgmt error code, should I do a mapping to a mgmt error code or leave it as an errno code and update the documentation?

de-nordic replied:
The SMP and errno should not overlap as SMP errors are positive and errnos are usually returned as negative.

Anywa,  I have proposed https://github.com/zephyrproject-rtos/zephyr/issues/42590 where I state my doubts whether "rc" should be used for errors not related to SMP processing, and here an error would not be related to it - because we have been able to process the SMP but failed at executing some operations. 
I was wandering whether in case of calculation/read/internal error the "rc", from SMP, should still be zero, but instead additional "err" code is logged where the system error is given.

lairdjm replied:
Fixed to use UNKNOWN error for now

de-nordic commented on subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/fs_mgmt.c:
```diff
@@ -208,11 +226,235 @@ fs_mgmt_file_upload(struct mgmt_ctxt *ctxt)
+
+	fs_close(&file);
+
+	/* Encode the response */
+	err = cbor_encode_text_stringz(&ctxt->encoder, "rc");
+	err |= cbor_encode_int(&ctxt->encoder, rc);
```
No, don't. We should stop using "rc" for reporting non SMP errors and success does not OK to be sent back.

lairdjm replied:
Done

de-nordic commented:
> @de-nordic As I understand you are accepting addition of functionalities in this PR. Am I right?

Yes. We have discussed them and I find them to be nice addition.

lairdjm commented on subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/hash_checksum_sha256.c:
```diff
@@ -0,0 +1,126 @@
+		rc = MGMT_ERR_EUNKNOWN;
+		goto error;
+	}
+#endif
```
Done

de-nordic commented on subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/fs_mgmt.c:
```diff
@@ -1,17 +1,43 @@
+#endif
+
+#ifdef CONFIG_FS_MGMT_CHECKSUM_HASH
+BUILD_ASSERT(sizeof(CONFIG_FS_MGMT_CHECKSUM_HASH_DEFAULT) > 1,
+	     "Missing mcumgr fs checksum/hash algorithm selection?");
+BUILD_ASSERT(CONFIG_FS_MGMT_CHECKSUM_HASH_LARGEST_OUTPUT_SIZE > 0,
+	     "Missing mcumgr fs checksum/hash algorithm selection?");
```
I am not sure whether this is needed; an user will not be able to set either of these with proj.conf because it is not allowed to set Kconfig options without prompt.

lairdjm replied:
They can enable the feature and not enable any algorithms for it which this catches

lairdjm commented on subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/fs_mgmt.c:
```diff
@@ -208,11 +234,227 @@ fs_mgmt_file_upload(struct mgmt_ctxt *ctxt)
+				tmp_val = (*(uint64_t *)output);
+			} else {
+				LOG_ERR("Unable to handle numerical checksum size %d",
+					group->output_size);
```
Fixed

de-nordic commented on subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/hash_checksum_mgmt.c:
```diff
@@ -0,0 +1,78 @@
+	}
+}
+
+static struct hash_checksum_mgmt_group *hash_checksum_mgmt_find_group(const char *name)
+{
+	struct hash_checksum_mgmt_group *group;
+
+	/* Find the group with the specified group name */
+	for (group = hash_checksum_mgmt_group_list; group != NULL; group = group->hc_mg_next) {
+		if (strcmp(group->group_name, name) == 0) {
+			break;
+		}
+	}
+
+	return group;
+}
```
Merge with `hash_checksum_mgmt_find_handler`, it seems that currently `hash_checksum_mgmt_find_handler` is only null check on return value from calling this.

lairdjm replied:
Fixed

de-nordic commented on subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/hash_checksum_mgmt.c:
```diff
@@ -0,0 +1,78 @@
+
+	group = hash_checksum_mgmt_find_group(name);
+	if (!group) {
+		return NULL;
```
I think that `return group` at line 77 is enough here.

lairdjm replied:
Fixed

de-nordic commented on subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/hash_checksum_mgmt.c:
```diff
@@ -0,0 +1,78 @@
+		return;
+	}
+
+	if (curr == group) {
+		hash_checksum_mgmt_group_list = curr->hc_mg_next;
+		return;
+	}
+
+	while (curr && curr != group) {
+		prev = curr;
+		curr = curr->hc_mg_next;
+	}
+
+	if (!prev || !curr) {
+		return;
+	}
+
+	prev->hc_mg_next = curr->hc_mg_next;
+	if (curr->hc_mg_next == NULL) {
+		hash_checksum_mgmt_group_list_end = curr;
+	}
```
You may use slist from Zephyr (https://docs.zephyrproject.org/latest/kernel/data_structures/slist.html) here.

lairdjm replied:
This the base code was a copy and paste from the mgmt handler functions, guess we need a bug to update that also

lairdjm replied:
FIxed

lairdjm commented on subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/fs_mgmt.c:
```diff
@@ -187,11 +213,217 @@ fs_mgmt_file_upload(struct mgmt_ctxt *ctxt)
+		     zcbor_int32_put(zse, rc);
+	} else {
+		ok = zcbor_tstr_put_lit(zse, "type")	&&
+		     zcbor_tstr_put_term(zse, type_arr);
```
`type` can only be used if the type was supplied, otherwise it will be empty as `type_arr` contains the default, unless you mean to do an if for which function to call and variable to use here?

de-nordic replied:
Right, I missed that. Leave it as is.

lairdjm commented on subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/Kconfig:
```diff
@@ -89,11 +89,65 @@ config FS_MGMT_DL_CHUNK_SIZE
+	default 32 if FS_MGMT_HASH_SHA256
+	default 4 if FS_MGMT_CHECKSUM_IEEE_CRC32
+	help
+	  Largest checksum/hash output size for internal buffer allocation.
```
Fixed

lairdjm commented on subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/include/fs_mgmt/hash_checksum_mgmt.h:
```diff
@@ -0,0 +1,80 @@
+/** @typedef hash_checksum_mgmt_handler_fn
+ * @brief Function that gets called to generate a hash or checksum.
+ *
+ * @param ctxt		mcumgr mgmt context
```
Fixed (was carry over from original PR where the hash function added data directly)

lairdjm commented on doc/services/device_mgmt/smp_groups/smp_group_8.rst:
```diff
@@ -216,3 +220,215 @@ where:
+
+In case when "rc" is not 0, success, the other fields will not appear.
+
+fuck
```
Good reminder to remove rc from the documentation, fixed.

lairdjm commented on subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/hash_checksum_mgmt.c:
```diff
@@ -0,0 +1,78 @@
+
+	if (!group) {
+		return;
+	}
```
Fixed

lairdjm commented on doc/releases/release-notes-3.1.rst:
```diff
@@ -187,6 +187,12 @@ Build and Infrastructure
+
+  * Added mcumgr fs handler for getting file status which returns file size
+  * Added mcumgr fs handler for getting file hash/checksum, with support for
+    IEEE CRC32 and SHA256
```
Updated

lairdjm commented on doc/releases/release-notes-3.1.rst:
```diff
@@ -187,6 +187,12 @@ Build and Infrastructure
 
+* Management
+
+  * Added mcumgr fs handler for getting file status which returns file size
```
Updated

de-nordic reviewed (approved):
Looks OK. Thanks!
